src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs

```diff
@@ -293,6 +433,136 @@ internal ProjectDependencyGraph WithAdditionalProjectReferences(ProjectId projec
+            {
+                if (!references.Contains(referencedProjectId))
+                {
+                    // If projectId references project, it isn't through referencedProjectId so the change doesn't
+                    // impact the dependency graph.
+                    //
+                    // Suppose we start with the following graph, and are removing the project reference A->B:
+                    //
+                    //   A -> B -> C
+                    //   A -> D
+                    //
+                    // This case is not hit for project C. The reverse transitive references for C contains B, which is
+                    // the target project of the removed reference. We can see that project C is impacted by this
+                    // removal, and after the removal, project A will no longer be in the reverse transitive references
+                    // of C.
+                    //
+                    // This case is hit for project D. The reverse transitive references for D does not contain B, which
+                    // means project D cannot be "downstream" of the impact of removing the reference A->B. We can see
+                    // that project A will still be in the reverse transitive references of D.
+                    //
+                    // This optimization does not catch all cases. For example, consider the following graph where we
+                    // are removing the project reference A->B:
+                    //
+                    //   A -> B -> D
+                    //   A -> D
+                    //
+                    // For this example, we do not hit this optimization because D contains project B in the set of
+                    // reverse transitive references. Without more complicated checks, we cannot rule out the
+                    // possibility that A may have been removed from the reverse transitive references of D by the
+                    // removal of the A->B reference.
+                    continue;
+                }
+
+                if (references.Contains(projectId))
+                {
+                    builder.Remove(project);
+                }
+            }
+
+            // Invalidate the transitive references from the previously-referenced project
+            builder.Remove(referencedProjectId);
+
+            return builder.ToImmutable();
+        }
+
+        internal ProjectDependencyGraph WithProjectRemoved(ProjectId projectId)
+        {
+            Contract.ThrowIfFalse(_projectIds.Contains(projectId));
+
+            // Project ID set and direct forward references are trivially updated by removing the key corresponding to
+            // the project getting removed.
+            var projectIds = _projectIds.Remove(projectId);
+            var referencesMap = ComputeNewReferencesMapForRemovedProject(_lazyReverseReferencesMap, _referencesMap, projectId);
```

Comment:
Consider named parameters here, especially since both _lazyReferenceReferencesMap and _referencesMap are the same type, so this could be reversed and break it. Doubly so since your method for the forward map and reverse map take things in opposite orders.

Reply:
Reordered for consistency and used named arguments locally; will mark resolved on push

Reply:
➡️ Now pushed.